### ISSUE TITLE
CTC loss does not converge when training confomer_base using 2 GPUS

### ISSUE THREAD

**Issue by louislau1129** [tencent-ailab/3m-asr#2]
Hi, thanks for the open-source code of 3m-asr, I am quite interested with this architecture and want to test it on Aishell1 first.
When I follow the run.sh, in the stage of training conformer_base model, I found CTC loss does not converge. From my point of view, the conformer_base is just a large version of conformer_embedding, however, the CTC loss converges smoothly in the latter case.

A different point between these two cases is that, I use 1 GPU to train conformer_embedding with batch_size=32 by default, and in the stage of conformer_base, the OOM issue arises, so 2 GPUs are used to train conformer_base with batch_size=16 in DDP mode in order to keep the same equal batch_size. 

![conformer_base_ctc](https://user-images.githubusercontent.com/45789277/164886081-6a471b78-0d45-46cb-b9b4-92a579390168.PNG)

**The loss figures are attached here (left is conformer_base, right is confrmer_embedding), any ideas? Also another question is that how about training the conformer_moe from scratch without pre-training conformer_embedding and conformer_base?** Thanks in advance.

**Comment by ododoyo** [comment]
Is your model trained on aishell1 dataset? If you switch to another datatset, the config may need some adapations.
From my point of view,  smaller `lr`, larger `warmup_steps` and larger `accum_grad` maybe helpful for the convergence problem. 
In our experiments, training conformer_moe from scratch without pretraining the conformer_base also has the problem that CTC loss does not converge. The conformer_embedding can be cancelled but it does give a better performance.

**Comment by louislau1129** [comment]
Thank you for your quick reply.
> Is your model trained on aishell1 dataset?

Yes, plus speed perturbation, about 350h data for training.


> smaller `lr`, larger `warmup_steps` and larger `accum_grad` maybe helpful

Ok, I will try it.

> The conformer_embedding can be cancelled

What do you mean by "cancel"? No pre-training or remove this module?

Regarding the CTC loss can converge very well on conformer_embedding but not well on the conformer_base, do you think there is any relationship with the DDP mode with 2 GPUs in the latter? or just because the conformer model becomes larger so the previous hyper params cannot suit well.

**Comment by ododoyo** [comment]
I mean conformer can be removed but it helps achieve a better performance. If you use this module, pre-training is also needed.

I don't recognize what is wrong with DDP. I think it's more likely caused by the training parameters.

**Comment by ododoyo** [comment]
I mean conformer_embedding can be removed but it helps achieve a better performance. If you use this module, pre-training is also needed.

I don't recognize what is wrong with DDP. I think it's more likely caused by the training parameters.

**Comment by louislau1129** [comment]
Thanks for your kind guidance! I finished the moe training, the CER on Aishell1 can be attained at 5.6%. I see it still lays behind to that reported CER (4.x%). May be it is related to the number of experts used, I only used 4 experts here. Do you agree this point?  Lastly, could you please point out where the "multi-level" implementation code is, it seems that I cannot find them or I got lost :(,

**Comment by ododoyo** [comment]
Sorry about the confusing results of aishell-1 testset. The reported CER(4.x%) on aishell-1 is  given by the model trained with the WenetSpeech dataset. Considering our MoE model with large capacity is designed for datasets of large scale, I haven't run our code on aishell-1 dataset. 

Generally, models with more experts can achieve better performance but the benefits may diminish with the increasing number of experts. I think the effect may also be affected when the training data is not sufficient. 

The implementation of ”multi-level“ is not provided yet since it didn't give a better performance than conformer_moe on the open WenetSpeech dataset in our experiments.

**Comment by louislau1129** [comment]
Ok, I see. Thanks!

**Comment by wlmsoft** [comment]
> Is your model trained on aishell1 dataset? If you switch to another datatset, the config may need some adapations. From my point of view, smaller `lr`, larger `warmup_steps` and larger `accum_grad` maybe helpful for the convergence problem. In our experiments, training conformer_moe from scratch without pretraining the conformer_base also has the problem that CTC loss does not converge. The conformer_embedding can be cancelled but it does give a better performance.

Thanks for the code! In this reply, you mean that the conformer_base is also should be pretrained during the moe training? 
Both the  non-expert part and expert part in conformer_moe should load the pretrained models ? If we have 4 experts, the  init   
 parameters of FFN-moe experts are same ?

Thanks!

**Comment by ododoyo** [comment]
Yes, in the given training pipeline, the conformer_moe model will load pretrained conformer_base model, including non-expert and expert part. The FFN modules in all experts are initialized the same with the FFN modules in conformer_base model

**Comment by wlmsoft** [comment]
> Yes, in the given training pipeline, the conformer_moe model will load pretrained conformer_base model, including non-expert and expert part. The FFN modules in all experts are initialized the same with the FFN modules in conformer_base model

Thanks!
